### PR TITLE
En 5949/hard fork process after import

### PR DESCRIPTION
**Pull request by sasurobert**
First PR of block and transaction processor after hardfork. Added TODO's for what is not resolved:

// TODO: do factory which creates the needed components
// TODO: consensus on the first created block -> integrate the hardfork after processor into the normal block processor ?
// TODO: save blocks to storage, marshalize, broadcast.
// TODO: set previous roothash to genesis block
// TODO: use new genesis block process
// TODO: unit and integration tests

### CODE REVIEW

File: update/genesis/import.go [hunk 1]
```diff
@@ -211,8 +232,57 @@ func newAccountCreator(accType Type) (state.AccountFactory, error) {
...
+		return trieForShard, nil
+	}
+
+	trieForShard, ok := si.tries[shardID]
+	if ok {
+		return trieForShard, nil
+	}
+
+	dbConfig := storageFactory.GetDBFromConfig(si.storageConfig.DB)
+	dbConfig.FilePath = path.Join(si.storageConfig.DB.FilePath, core.GetShardIdString(shardID))
```
Comment (iulianpascalau) [inline update/genesis/import.go:259]: we need to refactor this as it is very hard to test.

File: update/genesis/import.go [hunk 2]
```diff
@@ -4,38 +4,57 @@ import (
...
-	Hasher      hashing.Hasher
-	Marshalizer marshal.Marshalizer
+	Reader        update.MultiFileReader
+	Hasher        hashing.Hasher
+	Marshalizer   marshal.Marshalizer
+	ShardID       uint32
+	StorageConfig config.StorageConfig
 }
 
 type stateImport struct {
```
Comment (iulianpascalau) [inline update/genesis/import.go:36]: unit tests?

File: update/process/afterHardFork.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	mapBlockProcessors map[uint32]update.HardForkBlockProcessor
+	importHandler      update.ImportHandler
+	shardCoordinator   sharding.Coordinator
+	hasher             hashing.Hasher
+	marshalizer        marshal.Marshalizer
+}
+
+// NewAfterHardForkBlockCreation creates the after hard fork block creator process handler
+func NewAfterHardForkBlockCreation(args ArgsAfterHardFork) (*afterHardFork, error) {
+	if args.MapBlockProcessors != nil {
```
Comment (iulianpascalau) [inline update/process/afterHardFork.go:37]: check if nils?

File: update/process/afterHardFork.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package process
```
Comment (iulianpascalau) [inline update/process/afterHardFork.go:1]: unit tests?

File: update/process/metaBlock.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package process
```
Comment (iulianpascalau) [inline update/process/metaBlock.go:1]: unit tests?

File: update/process/processPending.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+	sort.Slice(sortedTxs, func(i, j int) bool {
+		return sortedTxs[i].hash < sortedTxs[j].hash
+	})
+
+	mapMiniBlocks := make(map[string]*block.MiniBlock)
+	for _, info := range sortedTxs {
+		rcvAddress, err := p.pubKeyConv.CreateAddressFromBytes(info.tx.GetRcvAddr())
+		if err != nil {
+			continue
```
Comment (iulianpascalau) [inline update/process/processPending.go:93]: might have some log.Debug lines here (all if error branches)

File: update/process/processPending.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		if dstShardID != p.shardCoordinator.SelfId() {
+			continue
+		}
+
+		blockType, err := p.processSingleTransaction(info)
+		if err != nil {
+			continue
+		}
+
+		localID := fmt.Sprintf("%d%d%d", sndShardId, dstShardID, blockType)
```
Comment (iulianpascalau) [inline update/process/processPending.go:113]: add some separators as this won't work as expected when sndShardId is 11 and dstShardID is 1
Suggestion: 
```GO
localID := fmt.Sprintf("%d_%d_%d", sndShardId, dstShardID, blockType)
```

File: update/process/processPending.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+			return miniBlocks[i].Type < miniBlocks[j].Type
+		}
+		return miniBlocks[i].SenderShardID < miniBlocks[j].SenderShardID
+	})
+
+	return miniBlocks, nil
+}
+
+func (p *pendingProcessor) processSingleTransaction(info *txInfo) (block.Type, error) {
+	rwdTx, ok := info.tx.(*rewardTx.RewardTx)
```
Comment (iulianpascalau) [inline update/process/processPending.go:148]: this is not quite extensible without modifying the code. I like the pattern found in the transaction coordinator that takes the pre/intermediate processor based on the tx block type and then send the transaction to that preprocessor.

File: update/process/processPending.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	Accounts         state.AccountsAdapter
+	TxProcessor      process.TransactionProcessor
+	RwdTxProcessor   process.RewardTransactionProcessor
+	ScrTxProcessor   process.SmartContractResultProcessor
+	PubKeyConv       state.PubkeyConverter
+	ShardCoordinator sharding.Coordinator
+	TxCoordinator    process.TransactionCoordinator
+}
+
+type pendingProcessor struct {
```
Comment (iulianpascalau) [inline update/process/processPending.go:30]: unit tests?

File: update/process/shardBlock.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+func (s *shardBlockCreator) createMiniBlockHeaders(body *block.Body) (int, []block.MiniBlockHeader, error) {
+	if len(body.MiniBlocks) == 0 {
+		return 0, nil, nil
+	}
+
+	totalTxCount := 0
+	miniBlockHeaders := make([]block.MiniBlockHeader, len(body.MiniBlocks))
+
+	for i := 0; i < len(body.MiniBlocks); i++ {
```
Comment (iulianpascalau) [inline update/process/shardBlock.go:149]: So we notarized the the newly generated shard headers inside the first metablock. We might implement this as well for genesis.

File: update/process/shardBlock.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+type ArgsNewShardBlockCreatorAfterHardFork struct {
+	ShardCoordinator   sharding.Coordinator
+	TxCoordinator      process.TransactionCoordinator
+	PendingTxProcessor update.PendingTransactionProcessor
+	ImportHandler      update.ImportHandler
+	Marshalizer        marshal.Marshalizer
+	Hasher             hashing.Hasher
+}
+
+type shardBlockCreator struct {
```
Comment (iulianpascalau) [inline update/process/shardBlock.go:27]: unit tests?

File: update/process/processPending.go (hunk 8 above)
Comment (sasurobert) [reply update/process/processPending.go:148]: these processors do not implement the same interface - preprocessors and postprocessors do much more then needed here. I will think about having a container plus same inteerface, but needs a lot of change in production code. Maybe as a separate task to make all these processors to implement the same ProcessTransaction interface ?

File: update/process/shardBlock.go (hunk 10 above)
Comment (sasurobert) [reply update/process/shardBlock.go:149]: yes. I thought it is better to save to storage and have all the stuff with it.

File: update/process/afterHardFork.go (hunk 3 above)
Comment (sasurobert) [reply update/process/afterHardFork.go:37]: that is a map that does not implement isinterfacenil.

File: update/genesis/import.go (hunk 2 above)
Comment (sasurobert) [reply update/genesis/import.go:36]: will be done in separate PR

File: update/process/afterHardFork.go (hunk 4 above)
Comment (sasurobert) [reply update/process/afterHardFork.go:1]: will be done in separate PR

File: update/process/metaBlock.go (hunk 5 above)
Comment (sasurobert) [reply update/process/metaBlock.go:1]: will be done in separate PR

File: update/process/processPending.go (hunk 9 above)
Comment (sasurobert) [reply update/process/processPending.go:30]: will be done in separate PR

File: update/process/shardBlock.go (hunk 11 above)
Comment (sasurobert) [reply update/process/shardBlock.go:27]: will be done in separate PR

File: update/genesis/import.go (hunk 1 above)
Comment (sasurobert) [reply update/genesis/import.go:259]: yes. it is hard to mock. but you have to create temporary persisters to import those tries which are not for your shard. How about creating a component which returns a trie storage manager according to a config, which implements and interface, so it is easy to mock ?

File: update/process/processPending.go (hunk 6 above)
Comment (sasurobert) [reply update/process/processPending.go:93]: done

File: update/process/processPending.go (hunk 7 above)
Comment (sasurobert) [reply update/process/processPending.go:113]: good catch

File: update/interface.go [hunk 12]
```diff
@@ -189,3 +212,8 @@ type SigVerifier interface {
 	Verify(message []byte, sig []byte, pk []byte) error
 	IsInterfaceNil() bool
 }
+
+// PendingTxProcessor defines the methods to process transactions after import
+type PendingTxProcessor interface {
```
Comment (SebastianMarian) [inline update/interface.go:217]: What is this interface for?

File: update/mock/trieFactoryStub.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
+package mock
+
+import (
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/data"
+)
+
+type TrieFactoryStub struct {
```
Comment (SebastianMarian) [inline update/mock/trieFactoryStub.go:8]: Add a mock comment

File: update/process/afterHardFork.go (hunk 3 above)
Comment (SebastianMarian) [inline update/process/afterHardFork.go:37]: == nil instead != nil for all these checks and also use check.IfNil where is possible

File: update/process/metaBlock.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/hashing"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+	"github.com/ElrondNetwork/elrond-go/update"
+)
+
+// ArgsNewMetaBlockCreatorAfterHardfork -
```
Comment (SebastianMarian) [inline update/process/metaBlock.go:16]: finish the comment

File: update/process/metaBlock.go [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+type metaBlockCreator struct {
+	importHandler    update.ImportHandler
+	marshalizer      marshal.Marshalizer
+	hasher           hashing.Hasher
+	shardCoordinator sharding.Coordinator
+}
+
+// NewMetaBlockCreatorAfterHardfork create the after hardfork metablock creator
```
Comment (SebastianMarian) [inline update/process/metaBlock.go:31]: creates

File: update/process/metaBlock.go [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	validatorRootHash, err := validatorAccounts.RootHash()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	accounts := m.importHandler.GetAccountsDBForShard(core.MetachainShardId)
+	if check.IfNil(accounts) {
+		return nil, nil, update.ErrNilAccounts
+	}
+	rootHash, err := m.importHandler.GetAccountsDBForShard(core.MetachainShardId).RootHash()
```
Comment (SebastianMarian) [inline update/process/metaBlock.go:75]: accounts.RootHash()

File: update/process/processPending.go [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+
+// ArgsPendingTransactionProcessor defines the arguments structure
+type ArgsPendingTransactionProcessor struct {
+	Accounts         state.AccountsAdapter
+	TxProcessor      process.TransactionProcessor
+	RwdTxProcessor   process.RewardTransactionProcessor
+	ScrTxProcessor   process.SmartContractResultProcessor
+	PubKeyConv       state.PubkeyConverter
+	ShardCoordinator sharding.Coordinator
+	TxCoordinator    process.TransactionCoordinator
```
Comment (SebastianMarian) [inline update/process/processPending.go:31]: not set or used

File: update/process/shardBlock.go [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+	if check.IfNil(args.ShardCoordinator) {
+		return nil, update.ErrNilShardCoordinator
+	}
+	if check.IfNil(args.TxCoordinator) {
+		return nil, update.ErrNilTxCoordinator
+	}
+	if check.IfNil(args.PendingTxProcessor) {
+		return nil, update.ErrNilPendingTxProcessor
+	}
+	if check.IfNil(args.ImportHandler) {
```
Comment (SebastianMarian) [inline update/process/shardBlock.go:56]: This check is already done in line 38. Remove it from here

File: update/mock/trieFactoryStub.go (hunk 13 above)
Comment (sasurobert) [reply update/mock/trieFactoryStub.go:8]: done.

File: update/interface.go (hunk 12 above)
Comment (sasurobert) [reply update/interface.go:217]: deleted.

File: update/genesis/import.go [hunk 19]
```diff
@@ -6,36 +6,53 @@ import (
...
+	validatorDB        state.AccountsAdapter
+	newRootHashes      map[uint32][]byte
+	validatorsRootHash []byte
+
+	hasher              hashing.Hasher
+	marshalizer         marshal.Marshalizer
+	storage             dataRetriever.StorageService
+	triesContainer      state.TriesHolder
+	trieStorageManagers map[string]data.StorageManager
+	shardID             uint32
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:50]: done

File: update/genesis/import.go [hunk 20]
```diff
@@ -6,36 +6,53 @@ import (
...
+	storageConfig       config.StorageConfig
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:51]: done

File: update/genesis/import.go [hunk 21]
```diff
@@ -6,36 +6,53 @@ import (
...
+	importedMetaBlock  *block.MetaBlock
+	tries              map[uint32]data.Trie
+	accountDBsMap      map[uint32]state.AccountsAdapter
+	validatorDB        state.AccountsAdapter
+	newRootHashes      map[uint32][]byte
+	validatorsRootHash []byte
+
+	hasher              hashing.Hasher
+	marshalizer         marshal.Marshalizer
+	storage             dataRetriever.StorageService
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:47]: done

File: update/genesis/import.go [hunk 22]
```diff
@@ -6,36 +6,53 @@ import (
...
+	triesContainer      state.TriesHolder
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:48]: done

File: update/genesis/import.go [hunk 23]
```diff
@@ -6,36 +6,53 @@ import (
...
+	trieStorageManagers map[string]data.StorageManager
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:49]: done

File: update/genesis/import.go [hunk 24]
```diff
@@ -211,8 +234,42 @@ func newAccountCreator(accType Type) (state.AccountFactory, error) {
...
+		return trieForShard, nil
+	}
+
+	trieForShard, ok := si.tries[shardID]
+	if ok {
+		return trieForShard, nil
+	}
+
+	var err error
+	_, trieForShard, err = si.trieFactory.Create(si.storageConfig, core.ShardIdToString(shardID), false)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply update/genesis/import.go:261]: done

File: update/process/afterHardFork.go (hunk 3 above)
Comment (sasurobert) [reply update/process/afterHardFork.go:37]: done

File: update/process/metaBlock.go (hunk 14 above)
Comment (sasurobert) [reply update/process/metaBlock.go:16]: done

File: update/process/metaBlock.go (hunk 15 above)
Comment (sasurobert) [reply update/process/metaBlock.go:31]: done

File: update/process/metaBlock.go (hunk 16 above)
Comment (sasurobert) [reply update/process/metaBlock.go:75]: done.

File: update/process/processPending.go (hunk 17 above)
Comment (sasurobert) [reply update/process/processPending.go:31]: done.

File: update/process/shardBlock.go (hunk 18 above)
Comment (sasurobert) [reply update/process/shardBlock.go:56]: done

File: update/genesis/import.go (hunk 1 above)
Comment (iulianpascalau) [reply update/genesis/import.go]: yup

File: update/process/afterHardFork.go [hunk 25]
```diff
@@ -34,7 +34,7 @@ type afterHardFork struct {
 
 // NewAfterHardForkBlockCreation creates the after hard fork block creator process handler
 func NewAfterHardForkBlockCreation(args ArgsAfterHardFork) (*afterHardFork, error) {
-	if args.MapBlockProcessors != nil {
+	if args.MapBlockProcessors == nil {
 		return nil, update.ErrNilHardForkBlockProcessor
 	}
 	if args.ImportHandler != nil {
```
Comment (SebastianMarian) [inline update/process/afterHardFork.go:40]: == nil or better check.IfNil instead != nil for all the checks in this method

Comment (sasurobert) [reply update/process/afterHardFork.go]: done

**Pull request merged**